### PR TITLE
fix: delete KIP finalizer and KIP operator on CheCluster CR delete

### PR DESCRIPTION
**Pull request by dkwon17**
Signed-off-by: David Kwon <dakwon@redhat.com>

Image: `quay.io/dkwon17/che-operator:cr_delete`

### What does this PR do?
Deleting the CheCluster CR when `spec.imagePuller.enable` is set to `true`, caused the CR to not be deleted, due to the Kubernetes Image Puller finalizer not being removed from the CR. This PR removes the finalizer and also removes the Kubernetes Image Puller Operator.

### Screenshot/screencast of this PR

Deleting the CheCluster CR when `spec.imagePuller.enable` is set to `true`, works properly.
The Kubernetes Image Puller Operator is also deleted.
cr-delete

### What issues does this PR fix or reference?

https://issues.redhat.com/browse/CRW-2256

### How to test this PR?

Tested on CRC.

1. Run the following to deploy Che:
```bash
chectl server:deploy -n eclipse-che --olm-channel=stable -p openshift --che-operator-image=quay.io/dkwon17/che-operator:cr_delete
```

2. After installation, set `spec.imagePuller.enable` is set to `true` in the CheCluster CR instance, and save changes:
image

3. Wait until the Kubernetes Image Puller Operator is installed:
image

4. Delete the CheCluster CR instance. The CR instance and the Image Puller Operator should both be deleted:
cr-delete

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [Custom resource definition file is up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#updating-custom-resource-definition-file)
- [ ] [OLM bundles are up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#update-olm-bundle)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: pkg/deploy/kubernetes_image_puller.go [hunk 1]
```diff
@@ -629,7 +643,8 @@ func UninstallImagePullerOperator(ctx *DeployContext) (bool, error) {
 	}
 
 	// Update CR to remove imagePullerSpec
-	if ctx.CheCluster.Spec.ImagePuller.Enable || ctx.CheCluster.Spec.ImagePuller.Spec != (chev1alpha1.KubernetesImagePullerSpec{}) {
+	if (ctx.CheCluster.Spec.ImagePuller.Enable || ctx.CheCluster.Spec.ImagePuller.Spec != (chev1alpha1.KubernetesImagePullerSpec{})) &&
```
Comment (tolusha) [inline pkg/deploy/kubernetes_image_puller.go:646]: Could you use `()` to emphasis the order of operations
`ctx.CheCluster.Spec.ImagePuller.Enable || ctx.CheCluster.Spec.ImagePuller.Spec != (chev1alpha1.KubernetesImagePullerSpec{})`

File: pkg/deploy/kubernetes_image_puller.go [hunk 2]
```diff
@@ -44,6 +44,20 @@ type ImageAndName struct {
 // a KubernetesImagePuller CR.  Add a finalizer to the CheCluster CR.  If false, remove the KubernetesImagePuller CR, uninstall the operator, and remove the finalizer.
 func ReconcileImagePuller(ctx *DeployContext) (reconcile.Result, error) {
 
+	if !ctx.CheCluster.ObjectMeta.DeletionTimestamp.IsZero() {
```
Comment (tolusha) [inline pkg/deploy/kubernetes_image_puller.go:47]: Could we avoid code duplication with: 
https://github.com/eclipse-che/che-operator/blob/main/pkg/deploy/kubernetes_image_puller.go#L173-L194

**Comment by dkwon17** [comment]
Thank you for the feedback, I have made changes. New image: `quay.io/dkwon17/che-operator:cr_delete_2`

File: pkg/deploy/kubernetes_image_puller.go [hunk 3]
```diff
@@ -171,27 +179,36 @@ func ReconcileImagePuller(ctx *DeployContext) (reconcile.Result, error) {
...
 			}
+		}
+	}
+	return reconcile.Result{}, nil
+}
 
-			if removed {
-				return reconcile.Result{RequeueAfter: time.Second}, nil
-			}
+func DeelteImagePullerOperatorAndFinalizer(ctx *DeployContext) (*reconcile.Result, error) {
```
Comment (tolusha) [inline pkg/deploy/kubernetes_image_puller.go:191]:
```suggestion
func DeleteImagePullerOperatorAndFinalizer(ctx *DeployContext) (*reconcile.Result, error) {
```

File: pkg/deploy/kubernetes_image_puller.go [hunk 4]
```diff
@@ -44,6 +44,14 @@ type ImageAndName struct {
 // a KubernetesImagePuller CR.  Add a finalizer to the CheCluster CR.  If false, remove the KubernetesImagePuller CR, uninstall the operator, and remove the finalizer.
 func ReconcileImagePuller(ctx *DeployContext) (reconcile.Result, error) {
 
+	if !ctx.CheCluster.ObjectMeta.DeletionTimestamp.IsZero() {
+		reconcileResult, err := DeelteImagePullerOperatorAndFinalizer(ctx)
```
Comment (tolusha) [inline pkg/deploy/kubernetes_image_puller.go:48]: It is better not to allow return nil.
Then we can do simply `return DeleteImagePullerOperatorAndFinalizer(ctx)`

File: pkg/deploy/kubernetes_image_puller.go [hunk 5]
```diff
@@ -171,27 +179,36 @@ func ReconcileImagePuller(ctx *DeployContext) (reconcile.Result, error) {
...
-				err = DeleteImagePullerFinalizer(ctx)
-				if err != nil {
-					logrus.Errorf("Error deleting finalizer: %v", err)
-					return reconcile.Result{}, err
-				}
-				return reconcile.Result{RequeueAfter: time.Second}, nil
-			}
+	if removed {
+		return &reconcile.Result{RequeueAfter: time.Second}, nil
+	}
```
Comment (tolusha) [inline pkg/deploy/kubernetes_image_puller.go:200]: Do we really need this?
That's not obvious.

Comment (dkwon17) [reply pkg/deploy/kubernetes_image_puller.go]: After some testing, it can be removed

**Review by tolusha — APPROVED** [review]

**Comment by tolusha** [comment]
pls. rebase on main

**Review by AndrienkoAleksandr — APPROVED** [review]

**Comment by dkwon17** [comment]
Sounds good, rebased with main

**Comment by dkwon17** [comment]
Updated with some fixes, however, I get this kind of error for the failing `ci/prow/v7` tests:
```
error: unable to parse image registry.build01.ci.openshift.org/ci-op-t6snlstl/stable@sha256:f08a109aeb8eb5e4cc81d075f5efe0bbf1ea25a5a67ff533d55b88723df5916a: cannot retrieve image configuration for manifest sha256:f08a109aeb8eb5e4cc81d075f5efe0bbf1ea25a5a67ff533d55b88723df5916a: received unexpected HTTP status: 500 Internal Server Error
{"component":"entrypoint","error":"wrapped process failed: exit status 1","file":"prow/entrypoint/run.go:80","func":"k8s.io/test-infra/prow/entrypoint.Options.Run","level":"error","msg":"Error executing test process","severity":"error","time":"2021-09-28T19:54:52Z"}
```
How can I fix this?

**Comment by AndrienkoAleksandr** [comment]
> How can I fix this?

Let's try one more time. Openshift ci tests isn't 100% stable. If it doesn't help, Openshift ci provides for us very a lot of different logs, to take a look them you can click "Artifacts" link.

**Comment by AndrienkoAleksandr** [comment]
LGTM, while You didn't faced with merge conflict.

**Review by tolusha — APPROVED** [review]

**Comment by dkwon17** [comment]
Thank you for the feedback everyone. If it is ok, could we merge soon?

**Pull request merged**